cpp/src/transform/transform.cpp

```diff
@@ -34,75 +34,173 @@ namespace cudf {
+                                             std::vector<column_view> const& inputs)
+{
+  static constexpr auto SCALAR_STRIDE = 0;
+  static constexpr auto COLUMN_STRIDE = 1;
+
+  auto const column_type_name = [](data_type data_type, bool is_scalar) {
+    return jitify2::reflection::Template("cudf::transformation::jit::strided")
+      .instantiate(type_to_name(data_type), is_scalar ? SCALAR_STRIDE : COLUMN_STRIDE);
+  };
+
+  std::vector<std::string> typenames;
+
+  typenames.push_back(column_type_name(output.type(), false));
+  std::transform(
+    inputs.begin(), inputs.end(), std::back_inserter(typenames), [&](auto const& input) {
+      bool const is_scalar = input.size() != output.size();
+      return column_type_name(input.type(), is_scalar);
+    });
+
+  return typenames;
+}
+
+std::map<unsigned int, std::string> build_ptx_params(mutable_column_view output,
+                                                     std::vector<column_view> const& inputs)
+{
+  std::map<unsigned int, std::string> params;
+  unsigned int index = 0;
+
+  auto const add_column = [&](bool is_output, data_type type) {
+    auto const param_type = type_to_name(type);
+    params.emplace(index++, is_output ? (param_type + "*") : param_type);
+  };
+
+  add_column(true, output.type());
+
+  for (auto& input : inputs) {
+    add_column(false, input.type());
+  }
+
+  return params;
+}
+
+std::vector<device_data_t> build_device_data(mutable_column_view output,
+                                             std::vector<column_view> const& inputs)
+{
+  std::vector<device_data_t> data;
+
+  data.push_back(const_cast<device_data_t>(cudf::jit::get_data_ptr(output)));
+
+  std::transform(inputs.begin(), inputs.end(), std::back_inserter(data), [](auto const& input) {
+    return const_cast<device_data_t>(cudf::jit::get_data_ptr(input));
+  });
+
+  return data;
+}
+
+std::vector<void*> build_launch_args(cudf::size_type& size, std::vector<device_data_t>& device_data)
 {
-  std::string const kernel_name =
-    jitify2::reflection::Template("cudf::transformation::jit::kernel")  //
-      .instantiate(cudf::type_to_name(output.type()),  // list of template arguments
-                   cudf::type_to_name(input.type()));
-
-  std::string cuda_source = is_ptx ? cudf::jit::parse_single_function_ptx(
-                                       udf,  //
-                                       "GENERIC_UNARY_OP",
-                                       {
-                                         {0, "void *"},                         // output argument
-                                         {1, cudf::type_to_name(input.type())}  // input argument
-                                       })
-                                   : cudf::jit::parse_single_function_cuda(udf,  //
-                                                                           "GENERIC_UNARY_OP");
+  // JITIFY and NVRTC need non-const pointers even if they aren't written
+  // to
```

Comment:
```suggestion
  // JITIFY and NVRTC need non-const pointers even if they are not modified
```

Reply:
But the phrase ends in a preposition :-) 
That's fine but can the `to` not be on a separate line on its own.